File: ios/Classes/Views/VideoTile/VideoTileView.swift
//
//  VideoTileView.swift
//  amazon_chime_plugin
//
//  Created by Yugo Sugiyama on 2023/08/29.
//

import AmazonChimeSDK
import AmazonChimeSDKMedia
import Foundation
import Flutter

final class VideoTileView: NSObject, FlutterPlatformView {
    private let videoRenderView: VideoRenderView = DefaultVideoRenderView()
    private let logger = ConsoleLogger(name: "MeetingSession")
    private let tileId: Int

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) {

        // Receieve tileId as a param.
        guard let params = args as? Dictionary<String, Any>,
              let tileId = params["tileId"] as? Int,
              let isMirror = params["isMirror"] as? Bool,
              let isScaleToFill = params["isScaleToFill"] as? Bool else {
            fatalError("`args` paramter must be `Int`!")
        }
        self.tileId = tileId
        super.init()

        // Memo: without delay, tileView won't present when re-display tileView
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            // Bind view to VideoView
            MeetingSession.shared.meetingSession?.audioVideo.bindVideoView(videoView: self.videoRenderView, tileId: tileId)
            // この辺りの設定をViewの宣言時に設定すると、なぜかUIが崩れるのでここに記載
            // Fix aspect ratio
            (self.videoRenderView as? UIView)?.contentMode = isScaleToFill ? .scaleToFill : .scaleAspectFit
            (self.videoRenderView as? DefaultVideoRenderView)?.mirror = isMirror

        }
        // Declare _view as UIView for Flutter interpretation
        //        videoRenderView = videoRenderView as UIView
    }

    deinit {
        // deinit when this view will removed from flutter widgets.
        MeetingSession.shared.meetingSession?.audioVideo.unbindVideoView(tileId: tileId)
    }

    func view() -> UIView {
        return videoRenderView as! UIView
    }
}

File: android/src/main/kotlin/com/example/amazon_chime_plugin/Views/VideoTile/VideoTileView.kt
package com.example.amazon_chime_plugin.Views.VideoTile

import android.content.Context
import android.os.Looper
import android.view.View
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.DefaultVideoRenderView
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.VideoScalingType
import com.amazonaws.services.chime.sdk.meetings.utils.logger.ConsoleLogger
import com.example.amazon_chime_plugin.Manager.MeetingSession
import io.flutter.plugin.platform.PlatformView

internal class VideoTileView(context: Context?, viewId: Int, creationParams: Map<String?, Any?>?): PlatformView {
    private val view: DefaultVideoRenderView
    private val tileId: Int

    private val videoTileViewLogger: ConsoleLogger = ConsoleLogger()

    override fun getView(): View {
        return view
    }

    override fun dispose() {
        MeetingSession.shared.meetingSession?.audioVideo?.unbindVideoView(tileId)
    }

    init {
        view = DefaultVideoRenderView(context as Context)
        val params = creationParams as Map<String, Any>
        tileId = params["tileId"] as Int
        val isScaleToFill = params["isScaleToFill"] as Boolean
        val isMirror = params["isMirror"] as Boolean
        view.scalingType = if(isScaleToFill) VideoScalingType.AspectFill else VideoScalingType.AspectFit
        view.mirror = isMirror
        // // Memo: without delay, tileView won't present when re-display tileView
        android.os.Handler(Looper.getMainLooper()).postDelayed( {
            MeetingSession.shared.meetingSession?.audioVideo?.bindVideoView(view, tileId)
                ?: videoTileViewLogger.error("VideoTileView", "Error while binding video view.")
        }, 300)
    }
}
